CRM/Core/Payment/BaseIPN.php

```diff
@@ -120,6 +112,40 @@ public function validateData(&$input, &$ids, &$objects, $required = TRUE, $payme
       echo "Failure: Could not find contribution record for {$contribution->id}<p>";
       return FALSE;
     }
+
+    // make sure contact exists and is valid
+    // use the contact id from the contribution record as the id in the IPN may not be valid anymore.
+    $contact = new CRM_Contact_BAO_Contact();
+    $contact->id = $contribution->contact_id;
+    if (!$contact->find(TRUE)) {
```

Comment:
To get here, what has to happen is that there must be a contribution in the database which is pointing to a contact which does not exist, in violation of database constraints.  The messages which appear below here don't seem to reflect this.  Unless I'm wrong here?

Reply:
@andrew-cormick-dockery well yes unless somehow you imported a contrib table with FK checks off and it had wonky data in it i guess but your probably right. One thing i just thought of should we test for if the contact is marked as is_deceased and error if it is? or is_deleted?

Reply:
Now that I look again at my pseudocode I can see that I made the same mistake myself.  It's clearly a fatal error if a contribution is referencing a contact ID that doesn't exist.  I don't see the need for testing multiple scenarios.

Now let's consider is_deleted.  This also violates rules, since you're not supposed to be able to delete contacts which do have contributions attached.  So it begs the question how did the record get into such a state in the first place.  So possibly, this is a fatal error (data model violation).

Now let's consider is_deceased.  This may actually be possible to do:

1. Person sets up recurring Paypal payment
2. Some time after this, the person sadly passes away
3. Spouse of late person informs the Greens, person is marked  as deceased.
4. What happens now with Paypal?!  I hope that the transaction got cancelled at step 3 because if it didn't, then Paypal could very well continue to process payments and send IPNs informing Civi  that it has done so.

In this case, the transaction very possibly occurred.  And it probably shouldn't have.  But it should be recorded, because it really did happen and it really needs to be fixed by an officer.

Reply:
I agree that any processed contribution should be recorded even in the situation that a contact may be marked as `is_deceased`.

As for `is_deleted` I argue we should still process the contribution. It _is_ possible to soft delete a contact which has contribution records. Hard deletion is likely also possible (though whether or not that's wise from a _business perspective_ is another discussion).